### PR TITLE
[MINI-5845] Support closing a MiniApp programmatically

### PR DESCRIPTION
**Pull request by GondiTilak**
# Description
Mini App should be able to close mini-app by calling the JS Interface and JS interface would call the native interface.

## Links
[MINI-5845](https://jira.rakuten-it.com/jira/browse/MINI-5845)
[Demo Video on BOX](https://rak.box.com/s/lupcyfphg9fod3qroz3xoha09y90s155)

# Checklist
- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `fastlane ci` without errors

### CODE REVIEW

File: Sources/Classes/core/Models/MiniAppJavaScriptMessageInfo.swift [hunk 1]
```diff
@@ -52,6 +53,10 @@ struct MiniAppCustomPermissionsListResponse: Codable {
     let status: String
 }
 
+struct MiniAppShouldClose: Decodable {
+    let withConfirmation: Bool?
+}
+
```
Comment (Climbatize) [inline Sources/Classes/core/Models/MiniAppJavaScriptMessageInfo.swift:59]: I'm not sure about the name of this struct, maybe something like `MiniAppCloseRequest`

File: Tests/Unit/Helpers.swift [hunk 2]
```diff
@@ -374,6 +374,10 @@ class MockFile {
 }
 
 class MockMessageInterfaceExtension: MiniAppMessageDelegate {
+    func closeMiniApp(withConfirmation: Bool, completionHandler: @escaping (Result<Bool, MASDKError>) -> Void) {
+        //TO-DO: implementation pending
```
Comment (Climbatize) [inline Tests/Unit/Helpers.swift:378]: Please don't commit TODOs, and name it without the dash: `//TODO: ` so they can be caught by analyse tools ;)

**Review by Climbatize — CHANGES_REQUESTED** [review]

File: Sources/Classes/core/Models/MiniAppJavaScriptMessageInfo.swift (hunk 1 above)
Comment (GondiTilak) [reply Sources/Classes/core/Models/MiniAppJavaScriptMessageInfo.swift]: This is not required hence removed it in latest commits.

File: Tests/Unit/Helpers.swift (hunk 2 above)
Comment (GondiTilak) [reply Tests/Unit/Helpers.swift]: Updated the test case implementation hence the TODO's are now removed.

File: Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift [hunk 3]
```diff
@@ -777,6 +779,24 @@ extension MiniAppScriptMessageHandler {
...
+        guard let requestParamValue = requestParam?.withConfirmationAlert else {
+            executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: prepareMAJavascriptError(MiniAppJavaScriptError.unexpectedMessageFormat))
+            return
+        }
+
+        self.hostAppMessageDelegate?.closeMiniApp(withConfirmation: requestParamValue, completionHandler: { result in
+            switch result {
+            case .success: ()
+            case .failure(let error):
+                self.handleMASDKError(error: error, callbackId: callbackId)
```
Comment (rleojoseph) [inline Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:794]: handleMASDKError is old method, do not use. You can use the following suggestion instead
```suggestion
     self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: prepareMAJavascriptError(error))

```

File: Sources/Classes/core/MiniAppAnalytics.swift [hunk 4]
```diff
@@ -38,6 +38,7 @@ internal enum MiniAppRATEvent: String, CaseIterable {
     case getSecureStorageSize = "mini_app_secure_storage_size"
     case setCloseAlertInfo = "mini_app_close_alert_info"
     case sendJsonToHostapp = "mini_app_send_json_to_host_app"
+    case closeMiniApp = "close_mini_app_with_confirmation_alert"
```
Comment (rleojoseph) [inline Sources/Classes/core/MiniAppAnalytics.swift:42]: Guess we are missing analytics for sendJsonToMiniapp method as well right ?

File: Example/Controllers/SwiftUI/Features/Single/MiniAppSingleView.swift [hunk 5]
```diff
@@ -168,6 +170,18 @@ struct MiniAppSingleView: View {
                 sendResumeMiniApp(miniAppId: viewModel.miniAppId, version: viewModel.miniAppVersion)
             }
         })
+        .onChange(of: viewModel.shouldCloseMiniApp.value, perform: { newValue in
+            if newValue {
+                if viewModel.closeWithConfirmation {
+                    closeAlertMessage = MiniAppAlertMessage(
+                        title: "Close Miniapp",
+                        message: "Are you sure, you want to close this MiniApp?"
+                    )
```
Comment (rleojoseph) [inline Example/Controllers/SwiftUI/Features/Single/MiniAppSingleView.swift:179]: You should not use the hardcoded close confirmation alert message here, rather you should the one which is set the mini-app itself using setCloseAlert

**Review by rleojoseph — CHANGES_REQUESTED** [review]
Overall looks good. Some changes required in the sample app. Please check

File: Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift (hunk 3 above)
Comment (GondiTilak) [reply Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift]: Updated the delegate method to use new `MiniAppJavaScriptError ` by making it public. And used the above suggested new  call back api call.

File: USERGUIDE.md [hunk 6]
```diff
@@ -1074,6 +1074,25 @@ window.addEventListener(MiniAppKeyboardEvents.KEYBOARDHIDDEN, function (e) {
...
+<a id="miniapp-close-event-listner></a>
+
+### MiniApp Close
+
+MiniApp can request the host app to close itself through the javascript bridge event trigger.
+MiniApp-SDK provides the `MiniAppMessageDelegate` interface method `func closeMiniApp(withConfirmation: , completionHandler:)` which must be implemented in the host app.
+
+```swift
+extension ViewController: MiniAppMessageDelegate {
+    func closeMiniApp(withConfirmation: Bool, completionHandler: @escaping (Result<Bool, MASDKError>) -> Void) {
```
Comment (rleojoseph) [inline USERGUIDE.md:1086]:
```suggestion
    func closeMiniApp(withConfirmation: Bool, completionHandler: @escaping (Result<Bool, MiniAppJavaScriptError >) -> Void) {
```

File: USERGUIDE.md [hunk 7]
```diff
@@ -1074,6 +1074,25 @@ window.addEventListener(MiniAppKeyboardEvents.KEYBOARDHIDDEN, function (e) {
 }
 ```
 
+<a id="miniapp-close-event-listner></a>
+
+### MiniApp Close
+
+MiniApp can request the host app to close itself through the javascript bridge event trigger.
+MiniApp-SDK provides the `MiniAppMessageDelegate` interface method `func closeMiniApp(withConfirmation: , completionHandler:)` which must be implemented in the host app.
```
Comment (rleojoseph) [inline USERGUIDE.md:1082]:
```suggestion
MiniApp iOS SDK provides the `MiniAppMessageDelegate` interface method `func closeMiniApp(withConfirmation: , completionHandler:)` which can be implemented in the host app.
```

Comment (rleojoseph) [reply USERGUIDE.md:1082]: Also, you can mention that this interface is optional

File: CHANGELOG.md [hunk 8]
```diff
@@ -3,6 +3,7 @@
 ### 5.1.0 (xxxx-xx-xx)
 **SDK**
 - **Update:** Added `allEmailList` field in `MAContact` to support multiple emails of a specific contact.
+- **Feature:** New `MiniAppMessageDelegate` interface method `func closeMiniApp(withConfirmation: , completionHandler:)` to handle the close miniapp action from MinApps.
```
Comment (rleojoseph) [inline CHANGELOG.md:6]:
```suggestion
- **Feature:** Added a new optional interface `func closeMiniApp(withConfirmation: , completionHandler:)` in  `MiniAppMessageDelegate` to support the close mini-app action from MinApps.
```

File: Sources/Classes/core/Protocols/MiniAppMessageDelegate.swift [hunk 9]
```diff
@@ -75,6 +78,10 @@ public extension MiniAppMessageDelegate {
     func downloadFile(fileName: String, url: String, headers: DownloadHeaders, completionHandler: @escaping (Result<String, MASDKError>) -> Void) {
         completionHandler(.failure(.failedToConformToProtocol))
     }
+
```
Comment (rleojoseph) [inline Sources/Classes/core/Protocols/MiniAppMessageDelegate.swift:81]: You forgot to implement 'closeMiniApp' interface

File: USERGUIDE.md [hunk 10]
```diff
@@ -1074,6 +1074,25 @@ window.addEventListener(MiniAppKeyboardEvents.KEYBOARDHIDDEN, function (e) {
 }
 ```
 
+<a id="miniapp-close-event-listner></a>
```
Comment (rleojoseph) [inline USERGUIDE.md:1077]:
```suggestion
<a id="miniapp-close></a>
```

File: USERGUIDE.md (hunk 7 above)
Comment (GondiTilak) [reply USERGUIDE.md]: Thank you I have updated in the latest commit.

File: CHANGELOG.md (hunk 8 above)
Comment (GondiTilak) [reply CHANGELOG.md]: Thank you I have updated in the latest commit.

File: Sources/Classes/core/Protocols/MiniAppMessageDelegate.swift (hunk 9 above)
Comment (GondiTilak) [reply Sources/Classes/core/Protocols/MiniAppMessageDelegate.swift:81]: Thank you for finding this, my bad I have updated in the latest commit.

File: USERGUIDE.md (hunk 10 above)
Comment (GondiTilak) [reply USERGUIDE.md]: Thank you I have updated in the latest commit.

File: USERGUIDE.md (hunk 6 above)
Comment (GondiTilak) [reply USERGUIDE.md]: Thank you I have updated in the latest commit.

File: Sources/Classes/core/MiniAppAnalytics.swift (hunk 4 above)
Comment (GondiTilak) [reply Sources/Classes/core/MiniAppAnalytics.swift:43]: Thank you I have updated in the latest commit.

File: Sources/Classes/core/MiniAppAnalytics.swift [hunk 11]
```diff
@@ -208,6 +212,8 @@ public class MiniAppAnalytics {
             return .setCloseAlertInfo
         case .sendJsonToHostapp:
             return .sendJsonToHostapp
```
Comment (rleojoseph) [inline Sources/Classes/core/MiniAppAnalytics.swift:214]: You may need to add for `sendJsonToMiniApp` ?

**Review by rleojoseph — APPROVED** [review]
Thank you for the changes. 👍🏻 . 
Looks good to me :+1:

File: Sources/Classes/core/MiniAppAnalytics.swift (hunk 11 above)
Comment (GondiTilak) [reply Sources/Classes/core/MiniAppAnalytics.swift:214]: We need not add it as it is not the `MiniAppJSActionCommand` type. This method in analytics is to track the JSAction commands received form the JS-Bridge I guess.

**Review by Climbatize — APPROVED** [review]

**Pull request merged**